### PR TITLE
[Skills Everywhere][#144][PORT] WaterfallHostBotDotNet to JavaScript

### PR DESCRIPTION
**Pull request by denscollo**
Addresses #144

## Description
This PR ports WaterfallHostBot DotNet to WaterfallHostBot JS using the following [sample](https://github.com/microsoft/BotFramework-FunctionalTests/tree/main/Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot).

## Testing
In the following image, there is a test run using WaterfallHostBotJS and WaterfallSkillBotJS.
image

### CODE REVIEW

**Comment by gabog** [comment]
Hi @joshgummersall, would you have some time to help reviewing this PR? (ping me offline if you need additional context)

File: Bots/JavaScript/Consumers/CodeFirst/WaterfallHostBot/authentication/allowedSkillsClaimsValidator.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+const { JwtTokenValidation, SkillValidation } = require('botframework-connector');
+const { SkillsConfiguration } = require('../skillsConfiguration');
+
+const skillsConfig = new SkillsConfiguration();
+// Load the appIds for the configured skills (we will only allow responses from skills we have configured).
+const allowedSkills = Object.values(skillsConfig.skills).map(skill => skill.appId);
```
Comment (joshgummersall) [inline Bots/JavaScript/Consumers/CodeFirst/WaterfallHostBot/authentication/allowedSkillsClaimsValidator.js:9]: I see this everywhere and I don't think it's necessary to fix it here, but going forward if we are looking something up at runtime this should be a `Set`, not an array.

File: Bots/JavaScript/Consumers/CodeFirst/WaterfallHostBot/bots/rootBot.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                // Run the Dialog with the activity.
+                await runDialog(this.mainDialog, turnContext, this.conversationState.createProperty('DialogState'));
+            }
+
+            await next();
+        });
+
+        this.onMembersAdded(async (context, next) => {
+            const membersAdded = context.activity.membersAdded;
+            for (let cnt = 0; cnt < membersAdded.length; ++cnt) {
```
Comment (joshgummersall) [inline Bots/JavaScript/Consumers/CodeFirst/WaterfallHostBot/bots/rootBot.js:38]: `for (const member of membersAdded) { ... }`

File: Bots/JavaScript/Consumers/CodeFirst/WaterfallHostBot/skillsConfiguration.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    get skillHostEndpoint() {
+        return this.skillHostEndpointValue;
+    }
+
+    createSkillDefinition(skill) {
+        // Note: we hard code this for now, we should dynamically create instances based on the manifests.
+        // For now, this code creates a strong typed version of the SkillDefinition and copies the info from
+        // settings into it.
+        let skillDefinition = new SkillDefinition();
```
Comment (joshgummersall) [inline Bots/JavaScript/Consumers/CodeFirst/WaterfallHostBot/skillsConfiguration.js:58]: I don't think the comment is accurate - to me, it looks like the `skillDefinition` is being overwritten below. `Object.assign(skillDefinition, new EchoSkill(), skill)` would (maybe) be more correct.

File: Bots/JavaScript/Consumers/CodeFirst/WaterfallHostBot/skillsConfiguration.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+        switch (true) {
```
Comment (joshgummersall) [inline Bots/JavaScript/Consumers/CodeFirst/WaterfallHostBot/skillsConfiguration.js:60]: This should be written as an `if-else`, I think that's more readable and won't require a bunch of break statements.

File: Bots/JavaScript/Consumers/CodeFirst/WaterfallHostBot/skillsConfiguration.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+const { SkillDefinition } = require('./skills/skillDefinition');
+
+/**
+ * A helper class that loads Skills information from configuration.
+ */
+class SkillsConfiguration {
+    constructor() {
+        this.skillsData = {};
+
+        var skillVariables = Object.keys(process.env).filter(prop => prop.startsWith('skill_'));
```
Comment (joshgummersall) [inline Bots/JavaScript/Consumers/CodeFirst/WaterfallHostBot/skillsConfiguration.js:16]: `var` should be `const` or `let`

I think functional style is easier to read and reason about, but this is just a suggestion. No need to change this if you disagree.

```javascript
this.skillsData = Object.entries(process.env).filter(([key]) => key.startsWith('skill_')).reduce((acc, [key, value]) => {
  const [, id, attr] = key.split('_');

  acc[id] = acc[id] || { definition: this.createSkillDefinition({ id }) };

  const propName = { appid: 'appId', endpoint: 'skillEndpoint' }[attr.toLowerCase()];
  if (!propName) { throw new Error(...) }

  acc[id][propName] = value;

  return acc;
}, {});
```

File: Bots/JavaScript/Consumers/CodeFirst/WaterfallHostBot/authentication/allowedSkillsClaimsValidator.js (hunk 1 above)
Comment (denscollo) [reply Bots/JavaScript/Consumers/CodeFirst/WaterfallHostBot/authentication/allowedSkillsClaimsValidator.js:9]: Thanks @joshgummersall , we'll keep this in mind to update all the JS bots soon.

**Review by gabog — APPROVED** [review]
Checked with @joshgummersall and this is good to go, I'll merge it as soon as the CI checks pass.

I'll also mark the test pipeline as non draft and will merge it tonight.

**Pull request merged**